### DISCUSSION TITLE
Strange evaluation of expression

### DISCUSSION THREAD

**Discussion by sirpi** [open-policy-agent/community#484] [Community]
Hi,

I created this rule:

```
getUnion = response {
  set1 := {"1", "2"}
  set2 := {"3", "4"}
  set3 := {"5", "6"}
  response = {"union": set1 | set2 | set3 }
}
```

And it returns 
{"union": {"1", "2"}} 
which is suprising.

If I wrap the sets with brackets: {"union": (set1 | set2 | set3) }
then the result is what we would expect (all six values)

I am pretty sure the issue is the precedence of operations. But what I don't understand is that I cannot reach the same result {"union": {"1", "2"}} using any brackets. Can someone tell me why I get this result? Should not this be some sort of error?

Note: one of the developers in our team found this while implementing a production policy which I presented in a simplified form.

**Comment by anderseknert** [comment]
Hi there! Heh, yeah that's a peculiar case :) What happens here is that the parser will see `{"union": set1 |` and say, "oh, that's a _map comprehension_,!" — a map comprehension where the the key is "union" and the value is the `set1` variable. `set2 | set3` will then be considered the body of the comprehension, but in this case the body will do nothing to contribute to the key/value of the map.

**Reply by anderseknert** [reply]
Docs on object comprehensions found here: https://www.openpolicyagent.org/docs/latest/policy-language/#object-comprehensions

**Reply by sirpi** [reply]
oh, thanks! Totally makes sense but I could not figure it out on my own. And I generate maps this way on a daily basis :)
